includes/class-wc-gateway-stripe.php

```diff
@@ -1136,22 +1148,111 @@ public function get_saved_payment_method_option_html( $token ) {
 	 * before displaying the Thank You page. This is useful in case webhooks have not been set up.
 	 *
 	 * @since 4.2.0
-	 *
 	 * @param int $order_id The ID that will be used for the thank you page.
 	 * @return int          The ID to use.
 	 */
 	public function thankyou_order_id( $order_id ) {
 		if ( $order_id > 0 ) {
 			$order = wc_get_order( $order_id );
-
-			if ( $order && 'pending' === $order->get_status() && $order->get_payment_method() === $this->id ) {
-				$intent = $this->get_intent_from_order( $order );
-				if ( 'succeeded' === $intent->status ) {
-					$this->process_response( end( $intent->charges->data ), $order );
-				}
-			}
+			$this->verify_intent_after_checkout( $order );
 		}
 
 		return $order_id;
 	}
+
+	/**
+	 * Attached to `woocommerce_payment_successful_result` with a late priority,
+	 * this method will combine the "naturally" generated redirect URL from
+	 * WooCommerce and a payment intent secret into a hash, which contains both
+	 * the secret, and a proper URL, which will confirm whether the intent succeeded.
+	 *
+	 * @since 4.2.0
+	 * @param array $result   The result from `process_payment`.
+	 * @param int   $order_id The ID of the order which is being paid for.
+	 * @return array
+	 */
+	public function modify_successful_payment_result( $result, $order_id ) {
+		// Only redirects with intents need to be modified.
+		if ( ! isset( $result['intent_secret'] ) ) {
+			return $result;
+		}
+
+		// Put the final thank you page redirect into the verification URL.
+		$verification_url = add_query_arg(
+			array(
+				'order'       => $order_id,
+				'nonce'       => wp_create_nonce( 'wc_stripe_confirm_pi' ),
+				'redirect_to' => rawurlencode( $result['redirect'] ),
+			),
+			WC_AJAX::get_endpoint( 'wc_stripe_verify_intent' )
+		);
+
+		// Combine into a hash.
+		$redirect = sprintf( '#confirm-pi-%s:%s', $result['intent_secret'], rawurlencode( $verification_url ) );
+
+		return array(
+			'result'   => 'success',
+			'redirect' => $redirect,
+		);
+	}
+
+	/**
+	 * Executed between the "Checkout" and "Thank you" pages, this
+	 * method updates orders based on the status of associated PaymentIntents.
+	 *
+	 * @since 4.2.0
+	 * @param WC_Order $order The order which is in a transitional state.
+	 */
+	public function verify_intent_after_checkout( $order ) {
+		if ( 'pending' !== $order->get_status() && 'failed' !== $order->get_status() ) {
```

Comment:
What's the point of executing the rest of this function if the order is in `failed` status? Do you think it's possible that the intent succeeds after the order was somehow marked as failed?

Reply:
Yes. If a customer submits the checkout form and gets to the SCA verification step, but decline, they remain on the checkout page (with an error message of course.

If they try to submit again and this time accept the SCA verification, the same order will be marked as paid/processing, instead of creating a new one.

Reply:
The same applies if they go to the "Pay for Order" screen of an order with a failed payment.

Reply:
Cool, thanks for the explanation 👍